### PR TITLE
Style: no empty newlines at start or end of functions

### PR DESCRIPTION
**Pull request by tkelman**
These were the few outliers style-wise that looked like this. Mostly from `@carlobaldassi`.

### CODE REVIEW

**Comment by amitmurthy** [comment]
Can we formalize this with an appropriate addition to https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#general-formatting-guidelines-for-julia-code-contributions

**Comment by nalimilan** [comment]
+1, we need to make the style guide more detailed.

**Comment by tkelman** [comment]
Sure, if there's agreement on this point then it's just a matter of how to phrase the guideline.

**Comment by amitmurthy** [comment]
Since source files are being changed to follow this style, may as well make it a guideline. I am fine with it.

**Comment by nalimilan** [comment]
See https://github.com/JuliaLang/julia/pull/15439.

**Comment by andreasnoack** [comment]
I think it can be useful for readability to have a newline in the beginning of a function when the signature is long. Is there a problem with allowing this? E.g. in terms of readability of diffs or something similar? Or is it mainly a matter of taste?

**Comment by tkelman** [comment]
Consistency. And it wastes space in a way that doesn't clarify anything in my opinion.

**Comment by tbreloff** [comment]
Personally, I like a space when the function can be broken up cleanly, or if there's a comment on the first line:

```julia
function nospace()
  dosomethingshort()
  dosomethingshort()
end

function yesspace()

  # first do this
  this()

  # then that
  that()

  # finally other
  other()

end
```

**Comment by tkelman** [comment]
This isn't unique to function definitions, any block opener/closer where you would typically indent the contents, having an empty newline right after the block opener or right before the `end` is redundant with the indentation and visually decouples the start and end of the block from its contents. It sticks out and doesn't look right to me. I'm all for whitespace between multiple steps, but before-start or after-finished don't count - that's what the `function` and the `end` are for.

**Comment by amitmurthy** [comment]
I concur with Tony on the consistency part, not just in a single file, but with rest of the codebase. Formalizing this as a preferred guideline (and not as a rule) would help. As Andreas points out, exceptions to the guideline could very well be long function signatures which wrap lines, where a newline would help with readability.

**Comment by nalimilan** [comment]
+1 for consistency as well.

File: base/bitarray.jl [hunk 1]
```diff
@@ -600,7 +600,6 @@ function insert!(B::BitVector, i::Integer, item)
 end
 
 function _deleteat!(B::BitVector, i::Integer)
-
     k, j = get_chunks_id(i)
```
Comment (samoconnor) [inline base/bitarray.jl]: Please don't delete blank lines in cases like this where the author has clearly taken some care to use whitespace to delineate functional blocks.

File: test/perf/kernel/stockcorr.jl [hunk 2]
```diff
@@ -4,7 +4,6 @@
 
 #STOCKCORR - The original, unoptimised code that simulates two correlated assets
 function stockcorr()
-
     ## Correlated asset information
```
Comment (samoconnor) [inline test/perf/kernel/stockcorr.jl]: Please don't delete blank lines preceding comments in cases like this where the author has clearly taken some time and care with manual ASCII layout to in an attempt to make the code easy to understand. A comment is the typographical equivalent of a heading, or the beginning of a paragraph, and should be clearly separated from whatever precedes it.

File: base/bitarray.jl (hunk 1 above)
Comment (tkelman) [inline base/bitarray.jl]: This makes the code less readable, as it detaches the contents of the function from the surrounding blocks that denote where the function starts and ends. a consistent policy is better than an inconsistent one, even if aesthetic judgements about it differ between contributors.

File: test/perf/kernel/stockcorr.jl (hunk 2 above)
Comment (tkelman) [inline test/perf/kernel/stockcorr.jl]: the extra line of separation here is completely redundant. the indentation already serves this exact purpose

**Pull request merged**